engineering/conventions.md

```diff
@@ -79,6 +80,18 @@ Your application will expose one or more binaries with a CLI.
 We prefer [GNU-style command line options](http://www.catb.org/esr/writings/taoup/html/ch10s05.html).
 That is, double dash for long options and single dash for short options (single letter). Include `--help` and `--version`. As a general rule, define always long flags, and then optionally short flags for tools that are often used interactively. When defining flags it is worth to consider matching behavior of widely used flags. For example, if `--all` is present, `-a` should be its shorthand. `--quiet` as a default instead of `--silent`. `--file/-f` for a file argument, `--output/-o` for output path, etc.
 
+## Logging
+
+* Use structured logging.
+* By default, output pretty logs when running in a terminal, output JSON when not.
+* Allow changing logging level and format explicitly through environment variables (`LOG_LEVEL`, `LOG_FORMAT`).
+* Use the following keys where they apply:
+  * `time` for log timestamp.
+  * `level` for log level.
+  * `error` for error message (optionally with stacktrace).
+  * `msg` for main log message.
+  * `source` for source file originating the log message (e.g. `foo.go:32`).
```

Comment:
I would also add `pid` if it not automatically appended - otherwise it can be hard to recover the exact event sequence on each node.

Reply:
I guess it might be useful in some scenarios, but note that Kubernetes appends a lot of data to each log entry, such as host, namespace, pod, etc.

Of course for applications with multithreading, additional data is added. For example, borges adds a worker ID to every log. In other cases you might add a session id to every log corresponding to a given request.

Reply:
Not requiring pid, since that will be handled by the log manager (e.g. Kubernetes in our case, systemd/journald or syslog for others). However, if you run multiple processes _inside a single container_ you might still want to add pid to your application. Otherwise, you won't be able to differentiate logs from different processes inside a single container.